chainer/functions/loss/discriminative_loss.py

```diff
@@ -0,0 +1,168 @@
+        self.delta_v = delta_v
+        self.delta_d = delta_d
+        self.alpha = alpha
+        self.beta = beta
+        self.gamma = gamma
+        self.max_embedding_dim = max_embedding_dim
+
+        if self.max_embedding_dim <= 0:
+            raise ValueError("Max number of embeddings has to be positive!")
+
+        # L1 or L2 norm is allowed only
+        if norm == 1:
+            self.norm = lambda x, axis=None: c_sum(absolute(x), axis=axis)
+        elif norm == 2:
+            self.norm = lambda x, axis=None: sqrt(c_sum(x ** 2, axis=axis))
+        else:
+            raise ValueError("For discriminative loss, "
+                             "norm can only be 1 or 2. "
+                             "Obtained the value : {}".format(norm))
+
+    def __call__(self, embeddings, labels):
+        """
+        :param embeddings: predicted embedding vectors
+        (batch size, max embedding dimensions, height, width)
+        :param labels: instance segmentation ground truth
+        each unique value has to be denoting one instance
+        (batch size, height, width)
+        :return:
+        tuple of chainer.Variable:
+            - Variance loss : variance loss multiplied by alpha
+            - Distance loss : distance loss multiplied by beta
+            - Regularization loss : regularization loss multiplied by gamma
+        """
+        assert (self.max_embedding_dim == embeddings.shape[1])
+
+        l_dist = 0.0
+        count = 0
+        xp = cuda.get_array_module(embeddings)
+
+        emb = embeddings[None, :]
+        emb = broadcast_to(emb, (emb.shape[1],
+                                 emb.shape[1],
+                                 emb.shape[2],
+                                 emb.shape[3],
+                                 emb.shape[4]))
+        ms = []
+        for c in range(self.max_embedding_dim):
+            # Create mask for instance
+            mask = xp.expand_dims(labels == c + 1, 1)
+            ms.append(mask)
+        ms = xp.stack(ms, 0)
+        mns = c_sum(emb * ms, axis=(3, 4))
+        mns = mns / xp.maximum(xp.sum(ms, (2, 3, 4))[:, :, None], 1)
+        mns_exp = mns[:, :, :, None, None]
+
+        # Calculate regularization term
+        l_reg = c_sum(self.norm(mns, (1, 2)))
+        l_reg = l_reg / (self.max_embedding_dim * embeddings.shape[0])
+
+        # Calculate variance term
+        l_var = self.norm((mns_exp - emb) * ms, 2)
+        l_var = relu(l_var - self.delta_v) ** 2
+        l_var = c_sum(l_var, (1, 2, 3))
+        l_var = l_var / xp.maximum(xp.sum(ms, (1, 2, 3, 4)), 1)
+        l_var = c_sum(l_var) / self.max_embedding_dim
+
+        # Calculate distance loss
+        for c_a in range(len(mns)):
+            for c_b in range(c_a + 1, len(mns)):
+                m_a = mns[c_a]
+                m_b = mns[c_b]
+                dist = self.norm(m_a - m_b, 1)  # N
+                l_dist += c_sum((relu(2 * self.delta_d - dist)) ** 2)
+                count += 1
```

Comment:
If we change the code so that computation of regularization and variance is free from for-loops, it would be better to change the computation of distance loss, too, because it takes O(N^2), which is dominant in computation time. But I also think it is ok that we change the implementation after we merge this PR.

Reply:
As I wrote in the previous PR, there is a trade-off between memory consumption and computation time. So, it is up to actual usecases which is better. Do you have any idea?

Reply:
Actually, I couldn't come up with a proper idea about dist term with matrix multiplication manner. We need to filter zero-means distance calculation (just noticed that current implementation has a bug about it). With matrix multiplication manner, we need some sort of mask to achieve it. Hence, I wonder if it will be much more efficient or not. So, I would like to keep it with current implementation as the current one is more like O(N * log(N)) rather than O(N ^ 2) - second for loop starts from the value of first for the loop -

Reply:
OK, I understand it. Let's reconsider the improvement when we encounter issues in current implementation.